Waciuma/tutorials (#772)
vmware-tanzu/cartographer: site/content/docs/development/tutorials/using-params.md

# Using Params

## Overview

In this tutorial we’ll explore how to use params to pass information that
isn’t anticipated by the standard workload fields. We’ll see how params can
either be set or delegated to other objects in the supply chain.

## Environment setup

For this tutorial you will need a kubernetes cluster with Cartographer
installed. You may follow
[the installation instructions here](https://github.com/vmware-tanzu/cartographer#installation).

Alternatively, you may choose to use the
[./hack/setup.sh](https://github.com/vmware-tanzu/cartographer/blob/main/hack/setup.sh)
script to install a kind cluster with Cartographer. _This script is meant for
our end-to-end testing and while we rely on it working in that role, no user
guarantees are made about the script._ Command to run from the Cartographer
directory:
```shell
$ ./hack/setup.sh cluster cartographer
```

If you later wish to tear down this generated cluster, run
```shell
$ ./hack/setup.sh teardown
```

## Scenario

> As in the tutorial "Build Your First Supply Chain", we will act as both the

Reviewer: ```suggestion
As in the tutorial ["Build Your First Supply Chain"](first-supply-chain.md), we will act as both the
```

and can make sure this object is in the expected namespaces with the
expected name. Let’s set this expected name as "expected-service-account".

While this will be the default value, we will allow the developer to
override this choice. We’ll do this by setting the name as a param for the
ClusterTemplate:

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterTemplate
metadata:
  ...
spec:
  template:
    ...
    spec:
      ...
      template:
        ...
        spec:
          serviceAccountName: $(params.image-pull-sa-name)$
          ...
  params:
    - name: image-pull-sa-name
      default: expected-service-account
```

Here we see two changes to the ClusterTemplate’s spec:
- The first is a field in the template. The field has been filled with a
  reference to a param: `$(params.image-pull-sa-name)$`
- Second, we see that a new field has been introduced to the spec: `params`.
  This is a list of params, each of which requires a name and a default
  value. Here, the app operator indicates that many devs will have a service
  account named `expected-service-account` in their namespace.

We apply this template to the cluster, along with the same supply chain from

> the "Your First Supply Chain" tutorial.

Reviewer 2: ```suggestion
the "Build Your First Supply Chain" tutorial.
```
might be also good to link to the supply chain, if possible

supply chain. Perhaps the template author is unsure of a parameter’s value,
but the supply chain author knows exactly the desired value and does not
want the workload author to be able to override their choice. The supply
chain can be altered to specify a `value`.

```yaml
---
apiVersion: carto.run/v1alpha1
kind: ClusterSupplyChain
metadata:
  name: supply-chain
spec:
  selector:
    workload-type: pre-built

  resources:
    - name: deploy
      templateRef:
        kind: ClusterTemplate
        name: app-deploy
      params:                                 # <=== New field
        - name: image-pull-sa-name
          value: inevitable-service-account
```

If the supply chain is redeployed with this definition, we can observe that
the deployment changes:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-deployment
  ...
spec:
  template:
    metadata: ...
    spec:
      Containers: ...
      dnsPolicy: ...
      restartPolicy: ...
      schedulerName: ...
      securityContext: {}
      serviceAccount: inevitable-service-account
      serviceAccountName: inevitable-service-account
      terminationGracePeriodSeconds: 30
  ...
status: ...
```

> Further information about params and the order of precedence can be found in
> Cartographer’s
> [RFC 10](https://github.com/vmware-tanzu/cartographer/blob/main/rfc/rfc-0010-supply-chains-supply-default-params.md).

Reviewer: ```suggestion
Further information about params and the order of precedence can be found in the ["Parameter Heirarchy"](../architecture/#parameter-hierarchy) architecture documentation.
```

In the Cartographer tests you can find an example of creating an object with
numerous parameters which demonstrates the precedence rules:
- [A template with many param fields, providing default values for some](https://github.com/vmware-tanzu/cartographer/tree/main/tests/kuttl/supplychain/params-supply-chain/00-proper-templates.yaml)
- [A supply chain providing some defaults and some values for params](https://github.com/vmware-tanzu/cartographer/tree/main/tests/kuttl/supplychain/params-supply-chain/01-supply-chain.yaml)
- [A workload providing some values for params](https://github.com/vmware-tanzu/cartographer/tree/main/tests/kuttl/supplychain/params-supply-chain/02-workload.yaml)
- [The expected object which will be created when that trio is submitted to
  the cluster](https://github.com/vmware-tanzu/cartographer/tree/main/tests/kuttl/supplychain/params-supply-chain/02-assert.yaml)

> ## Wrapup

Reviewer 2: nit
```suggestion
## Wrap Up
```
